File: resources/.pico8ide-skills/p8mod-author/scripts/validate-p8mod.py
#!/usr/bin/env python3
"""Validate Pico-8 IDE .p8mod structure.

Usage:
  python3 .codex/skills/p8mod-author/scripts/validate-p8mod.py path/to/game.p8mod
"""

from __future__ import annotations

import json
import re
import sys
from pathlib import Path

HEADER = "pico-8 cartridge // http://www.pico-8.com"
STANDARD_SECTIONS = ["lua", "gfx", "gff", "map", "sfx", "music"]
CUSTOM_SECTIONS = ["meta", "i18n"]
SECTION_RE = re.compile(r"^__(\w+)__$")


def fail(errors: list[str], message: str) -> None:
    errors.append(f"ERROR: {message}")


def warn(warnings: list[str], message: str) -> None:
    warnings.append(f"WARNING: {message}")


def parse_sections(lines: list[str]) -> tuple[dict[str, list[str]], list[tuple[str, int]], list[str]]:
    sections: dict[str, list[str]] = {}
    order: list[tuple[str, int]] = []
    duplicates: list[str] = []
    current: str | None = None

    for line_no, line in enumerate(lines, start=1):
        match = SECTION_RE.match(line)
        if match:
            current = match.group(1)
            if current in sections:
                duplicates.append(current)
            sections[current] = []
            order.append((current, line_no))
        elif current is not None:
            sections[current].append(line)

    return sections, order, duplicates


def validate_json_object(name: str, body: str, errors: list[str]) -> object | None:
    if not body.strip():
        fail(errors, f"__{name}__ section is empty")
        return None
    try:
        parsed = json.loads(body)
    except json.JSONDecodeError as exc:
        fail(errors, f"__{name}__ invalid JSON: {exc}")
        return None
    if not isinstance(parsed, dict):
        fail(errors, f"__{name}__ must be a JSON object")
        return None
    return parsed


def validate_meta(meta: object, errors: list[str]) -> None:
    if not isinstance(meta, dict):
        return
    for key in ("title", "author", "template"):
        if key in meta and not isinstance(meta[key], str):
            fail(errors, f"__meta__.{key} must be a string")


def validate_i18n(i18n: object, lua_body: str, errors: list[str], warnings: list[str]) -> None:
    if not isinstance(i18n, dict):
        return

    locales = i18n.get("locales")
    entries = i18n.get("entries")
    output_locale = i18n.get("outputLocale", "")

    if not isinstance(locales, list) or not all(isinstance(loc, str) and loc for loc in locales):
        fail(errors, "__i18n__.locales must be an array of non-empty strings")
        locales = []
    if not isinstance(entries, list):
        fail(errors, "__i18n__.entries must be an array")
        entries = []
    if not isinstance(output_locale, str):
        fail(errors, "__i18n__.outputLocale must be a string")
    elif output_locale and output_locale not in locales:
        fail(errors, f'__i18n__.outputLocale "{output_locale}" is not in locales')

    seen_keys: set[str] = set()
    for idx, entry in enumerate(entries):
        if not isinstance(entry, dict):
            fail(errors, f"__i18n__.entries[{idx}] must be an object")
            continue
        key = entry.get("key")
        translations = entry.get("translations")
        if not isinstance(key, str) or not key:
            fail(errors, f"__i18n__.entries[{idx}].key must be a non-empty string")
            continue
        if key in seen_keys:
            fail(errors, f'duplicate i18n key "{key}"')
        seen_keys.add(key)
        if not isinstance(translations, dict):
            fail(errors, f'entry "{key}" translations must be an object')
            continue
        for loc in locales:
            value = translations.get(loc)
            if not isinstance(value, str):
                fail(errors, f'entry "{key}" missing string translation for locale "{loc}"')

    tx_keys = set(re.findall(r'tx\(\s*["\']([^"\']+)["\']', lua_body))
    for key in sorted(tx_keys - seen_keys):
        fail(errors, f'Lua calls tx("{key}") but __i18n__ has no matching entry')
    for key in sorted(seen_keys - tx_keys):
        warn(warnings, f'__i18n__ entry "{key}" is not referenced by tx() in __lua__')
    if tx_keys and "_txi()" not in lua_body:
        warn(warnings, "Lua uses tx() but _txi() was not found")


def validate(path: Path) -> int:
    errors: list[str] = []
    warnings: list[str] = []

    try:
        text = path.read_text(encoding="utf-8")
    except UnicodeDecodeError:
        fail(errors, "file is not valid UTF-8 text")
        text = ""

    lines = text.splitlines()
    if not lines or lines[0] != HEADER:
        fail(errors, f"first line must be exactly: {HEADER}")
    if len(lines) < 2 or lines[1] != "version 42":
        fail(errors, "second line must be exactly: version 42")
    if text and not text.endswith("\n"):
        warn(warnings, "file should end with a trailing newline")

    sections, order, duplicates = parse_sections(lines)
    for duplicate in duplicates:
        fail(errors, f"duplicate section __{duplicate}__")
    for section in STANDARD_SECTIONS:
        if section not in sections:
            fail(errors, f"missing required section __{section}__")

    known = set(STANDARD_SECTIONS + CUSTOM_SECTIONS)
    for section, line_no in order:
        if section not in known:
            warn(warnings, f"unknown section __{section}__ at line {line_no}")

    section_names = [name for name, _ in order]
    expected_standard = [name for name in STANDARD_SECTIONS if name in sections]
    actual_standard = [name for name in section_names if name in STANDARD_SECTIONS]
    if actual_standard != expected_standard:
        fail(errors, "standard sections must appear in order: " + ", ".join(f"__{s}__" for s in STANDARD_SECTIONS))
    if any(name in CUSTOM_SECTIONS for name in section_names):
        first_custom = min(i for i, name in enumerate(section_names) if name in CUSTOM_SECTIONS)
        if any(name in STANDARD_SECTIONS for name in section_names[first_custom + 1 :]):
            fail(errors, "custom sections must appear after all standard sections")

    gfx = sections.get("gfx", [])
    if len(gfx) != 128:
        fail(errors, f"__gfx__ must contain 128 lines, found {len(gfx)}")
    for idx, line in enumerate(gfx, start=1):
        if not re.fullmatch(r"[0-9a-f]{128}", line):
            fail(errors, f"__gfx__ line {idx} must be 128 lowercase hex characters")
            break

    gff = sections.get("gff", [])
    if len(gff) != 2:
        fail(errors, f"__gff__ must contain 2 lines, found {len(gff)}")
    for idx, line in enumerate(gff, start=1):
        if not re.fullmatch(r"[0-9a-f]{256}", line):
            fail(errors, f"__gff__ line {idx} must be 256 lowercase hex characters")
            break

    for idx, line in enumerate(sections.get("map", []), start=1):
        if not re.fullmatch(r"[0-9a-f]{256}", line):
            fail(errors, f"__map__ line {idx} must be 256 lowercase hex characters")
            break
    if len(sections.get("map", [])) > 32:
        fail(errors, "__map__ must not contain more than 32 lines")

    for idx, line in enumerate(sections.get("sfx", []), start=1):
        if not re.fullmatch(r"[0-9a-f]{168}", line):
            fail(errors, f"__sfx__ line {idx} must be 168 lowercase hex characters")
            break
    if len(sections.get("sfx", [])) > 64:
        fail(errors, "__sfx__ must not contain more than 64 lines")

    for idx, line in enumerate(sections.get("music", []), start=1):
        if not re.fullmatch(r"[0-9a-f]{2} [0-9a-f]{8}", line):
            fail(errors, f"__music__ line {idx} must match 'ff cccccccc' lowercase hex format")
            break
    if len(sections.get("music", [])) > 64:
        fail(errors, "__music__ must not contain more than 64 lines")

    if "meta" in sections:
        meta = validate_json_object("meta", "\n".join(sections["meta"]), errors)
        validate_meta(meta, errors)
    if "i18n" in sections:
        i18n = validate_json_object("i18n", "\n".join(sections["i18n"]), errors)
        validate_i18n(i18n, "\n".join(sections.get("lua", [])), errors, warnings)

    for message in warnings:
        print(message)
    for message in errors:
        print(message)

    if errors:
        print(f"FAILED: {path} has {len(errors)} error(s)")
        return 1
    print(f"OK: {path} is a structurally valid .p8mod file")
    return 0


def main(argv: list[str]) -> int:
    if len(argv) != 2:
        print(__doc__.strip())
        return 2
    path = Path(argv[1])
    if not path.is_file():
        print(f"ERROR: file not found: {path}")
        return 1
    return validate(path)


if __name__ == "__main__":
    raise SystemExit(main(sys.argv))

File: resources/.pico8ide-skills/p8mod-author/scripts/validate-p8mod.mjs
#!/usr/bin/env node
// Validate Pico-8 IDE .p8mod structure.
// Usage: node .codex/skills/p8mod-author/scripts/validate-p8mod.mjs path/to/game.p8mod

import fs from 'node:fs';

const HEADER = 'pico-8 cartridge // http://www.pico-8.com';
const STANDARD_SECTIONS = ['lua', 'gfx', 'gff', 'map', 'sfx', 'music'];
const CUSTOM_SECTIONS = ['meta', 'i18n'];
const SECTION_RE = /^__(\w+)__$/;

function usage() {
  console.log('Usage: node .codex/skills/p8mod-author/scripts/validate-p8mod.mjs path/to/game.p8mod');
}

function parseSections(lines) {
  const sections = new Map();
  const order = [];
  const duplicates = [];
  let current = null;

  lines.forEach((line, index) => {
    const match = line.match(SECTION_RE);
    if (match) {
      current = match[1];
      if (sections.has(current)) duplicates.push(current);
      sections.set(current, []);
      order.push([current, index + 1]);
    } else if (current) {
      sections.get(current).push(line);
    }
  });

  return { sections, order, duplicates };
}

function parseJsonSection(name, lines, errors) {
  const body = lines.join('\n').trim();
  if (!body) {
    errors.push(`ERROR: __${name}__ section is empty`);
    return null;
  }
  try {
    const parsed = JSON.parse(body);
    if (!parsed || Array.isArray(parsed) || typeof parsed !== 'object') {
      errors.push(`ERROR: __${name}__ must be a JSON object`);
      return null;
    }
    return parsed;
  } catch (error) {
    errors.push(`ERROR: __${name}__ invalid JSON: ${error.message}`);
    return null;
  }
}

function validateMeta(meta, errors) {
  if (!meta) return;
  for (const key of ['title', 'author', 'template']) {
    if (key in meta && typeof meta[key] !== 'string') {
      errors.push(`ERROR: __meta__.${key} must be a string`);
    }
  }
}

function validateI18n(i18n, luaBody, errors, warnings) {
  if (!i18n) return;
  let locales = i18n.locales;
  let entries = i18n.entries;
  const outputLocale = i18n.outputLocale ?? '';

  if (!Array.isArray(locales) || !locales.every((loc) => typeof loc === 'string' && loc.length > 0)) {
    errors.push('ERROR: __i18n__.locales must be an array of non-empty strings');
    locales = [];
  }
  if (!Array.isArray(entries)) {
    errors.push('ERROR: __i18n__.entries must be an array');
    entries = [];
  }
  if (typeof outputLocale !== 'string') {
    errors.push('ERROR: __i18n__.outputLocale must be a string');
  } else if (outputLocale && !locales.includes(outputLocale)) {
    errors.push(`ERROR: __i18n__.outputLocale "${outputLocale}" is not in locales`);
  }

  const seenKeys = new Set();
  entries.forEach((entry, index) => {
    if (!entry || Array.isArray(entry) || typeof entry !== 'object') {
      errors.push(`ERROR: __i18n__.entries[${index}] must be an object`);
      return;
    }
    const key = entry.key;
    const translations = entry.translations;
    if (typeof key !== 'string' || !key) {
      errors.push(`ERROR: __i18n__.entries[${index}].key must be a non-empty string`);
      return;
    }
    if (seenKeys.has(key)) errors.push(`ERROR: duplicate i18n key "${key}"`);
    seenKeys.add(key);
    if (!translations || Array.isArray(translations) || typeof translations !== 'object') {
      errors.push(`ERROR: entry "${key}" translations must be an object`);
      return;
    }
    for (const loc of locales) {
      if (typeof translations[loc] !== 'string') {
        errors.push(`ERROR: entry "${key}" missing string translation for locale "${loc}"`);
      }
    }
  });

  const txKeys = new Set([...luaBody.matchAll(/tx\(\s*["']([^"']+)["']/g)].map((m) => m[1]));
  for (const key of [...txKeys].sort()) {
    if (!seenKeys.has(key)) errors.push(`ERROR: Lua calls tx("${key}") but __i18n__ has no matching entry`);
  }
  for (const key of [...seenKeys].sort()) {
    if (!txKeys.has(key)) warnings.push(`WARNING: __i18n__ entry "${key}" is not referenced by tx() in __lua__`);
  }
  if (txKeys.size > 0 && !luaBody.includes('_txi()')) {
    warnings.push('WARNING: Lua uses tx() but _txi() was not found');
  }
}

function validateHexLines(sectionName, lines, expectedCount, expectedPattern, errors) {
  if (expectedCount !== null && lines.length !== expectedCount) {
    errors.push(`ERROR: __${sectionName}__ must contain ${expectedCount} lines, found ${lines.length}`);
  }
  lines.some((line, index) => {
    if (!expectedPattern.test(line)) {
      errors.push(`ERROR: __${sectionName}__ line ${index + 1} has invalid format`);
      return true;
    }
    return false;
  });
}

function validate(path) {
  const errors = [];
  const warnings = [];
  let text;
  try {
    text = fs.readFileSync(path, 'utf8');
  } catch (error) {
    console.log(`ERROR: cannot read ${path}: ${error.message}`);
    return 1;
  }

  const lines = text.split(/\r?\n/);
  if (lines.at(-1) === '') lines.pop();
  if (lines[0] !== HEADER) errors.push(`ERROR: first line must be exactly: ${HEADER}`);
  if (lines[1] !== 'version 42') errors.push('ERROR: second line must be exactly: version 42');
  if (text && !text.endsWith('\n')) warnings.push('WARNING: file should end with a trailing newline');

  const { sections, order, duplicates } = parseSections(lines);
  for (const duplicate of duplicates) errors.push(`ERROR: duplicate section __${duplicate}__`);
  for (const section of STANDARD_SECTIONS) {
    if (!sections.has(section)) errors.push(`ERROR: missing required section __${section}__`);
  }

  const known = new Set([...STANDARD_SECTIONS, ...CUSTOM_SECTIONS]);
  for (const [section, lineNo] of order) {
    if (!known.has(section)) warnings.push(`WARNING: unknown section __${section}__ at line ${lineNo}`);
  }

  const sectionNames = order.map(([name]) => name);
  const actualStandard = sectionNames.filter((name) => STANDARD_SECTIONS.includes(name));
  const expectedStandard = STANDARD_SECTIONS.filter((name) => sections.has(name));
  if (actualStandard.join('|') !== expectedStandard.join('|')) {
    errors.push(`ERROR: standard sections must appear in order: ${STANDARD_SECTIONS.map((s) => `__${s}__`).join(', ')}`);
  }
  const firstCustom = sectionNames.findIndex((name) => CUSTOM_SECTIONS.includes(name));
  if (firstCustom !== -1 && sectionNames.slice(firstCustom + 1).some((name) => STANDARD_SECTIONS.includes(name))) {
    errors.push('ERROR: custom sections must appear after all standard sections');
  }

  validateHexLines('gfx', sections.get('gfx') ?? [], 128, /^[0-9a-f]{128}$/, errors);
  validateHexLines('gff', sections.get('gff') ?? [], 2, /^[0-9a-f]{256}$/, errors);

  const mapLines = sections.get('map') ?? [];
  if (mapLines.length > 32) errors.push('ERROR: __map__ must not contain more than 32 lines');
  validateHexLines('map', mapLines, null, /^[0-9a-f]{256}$/, errors);

  const sfxLines = sections.get('sfx') ?? [];
  if (sfxLines.length > 64) errors.push('ERROR: __sfx__ must not contain more than 64 lines');
  validateHexLines('sfx', sfxLines, null, /^[0-9a-f]{168}$/, errors);

  const musicLines = sections.get('music') ?? [];
  if (musicLines.length > 64) errors.push('ERROR: __music__ must not contain more than 64 lines');
  validateHexLines('music', musicLines, null, /^[0-9a-f]{2} [0-9a-f]{8}$/, errors);

  if (sections.has('meta')) validateMeta(parseJsonSection('meta', sections.get('meta'), errors), errors);
  if (sections.has('i18n')) {
    validateI18n(parseJsonSection('i18n', sections.get('i18n'), errors), (sections.get('lua') ?? []).join('\n'), errors, warnings);
  }

  for (const message of warnings) console.log(message);
  for (const message of errors) console.log(message);
  if (errors.length > 0) {
    console.log(`FAILED: ${path} has ${errors.length} error(s)`);
    return 1;
  }
  console.log(`OK: ${path} is a structurally valid .p8mod file`);
  return 0;
}

if (process.argv.length !== 3) {
  usage();
  process.exit(2);
}

if (!fs.existsSync(process.argv[2]) || !fs.statSync(process.argv[2]).isFile()) {
  console.log(`ERROR: file not found: ${process.argv[2]}`);
  process.exit(1);
}

process.exit(validate(process.argv[2]));
